src/server/gc_worker/compaction_filter.rs

```diff
@@ -699,13 +701,22 @@ fn check_need_gc(
             return (true, false);
         }
 
-        // When comparing `num_versions` with `num_puts`, trait internal levels
-        // specially because MVCC-deletion marks can't be handled at those
-        // levels.
-        let num_rollback_and_locks = (props.num_versions - props.num_deletes) as f64;
-        if num_rollback_and_locks > props.num_puts as f64 * ratio_threshold {
+        if key_mode == KeyMode::Txn || key_mode == KeyMode::Tidb {
+            // When comparing `num_versions` with `num_puts`, trait internal levels
+            // specially because MVCC-deletion marks can't be handled at those
+            // levels.
+            let num_rollback_and_locks = (props.num_versions - props.num_deletes) as f64;
+            if num_rollback_and_locks > props.num_puts as f64 * ratio_threshold {
+                return (true, false);
+            }
+        } else if props.num_deletes as f64 > props.num_puts as f64 * ratio_threshold {
             return (true, false);
```

Comment:
We handle GC of deletion at bottom most level only. So we should ignore `num_deletes` here ?

And according to the codes of Txn above, `num_deletes` is also considered to be no need to GC in Txn.